### PR TITLE
[various] Switch from glob to fast-glob

### PR DESCRIPTION
**Pull request by dmichon-msft**
## Summary
Switches from `glob` to `fast-glob` for globbing. The latter ships built-in typings, has a promise-based API, and leverages the `withFileTypes: true` option of `readdir` for better performance.

## Details
Affects 3 packages.

## How it was tested
Full build and test.

## Impacted documentation
None.

### CODE REVIEW

File: libraries/typings-generator/src/TypingsGenerator.ts [hunk 1]
```diff
@@ -134,11 +133,10 @@ export class TypingsGenerator {
     let checkFilePaths: boolean = true;
     if (!relativeFilePaths?.length) {
       checkFilePaths = false; // Don't check file paths if we generate them
-      relativeFilePaths = await LegacyAdapters.convertCallbackToPromise(glob, this.inputFileGlob, {
+      relativeFilePaths = await glob(this.inputFileGlob, {
         cwd: this.sourceFolderPath,
-        ignore: this.ignoredFileGlobs,
-        nosort: true,
-        nodir: true
+        ignore: this.ignoredFileGlobs as string[],
```
Comment (D4N14L) [inline libraries/typings-generator/src/TypingsGenerator.ts:138]: Hmm, should we allocate a new array for this? It's unlikely it'll get changed so this is not really a concern of mine, but more asking for my own knowledge of what the best design here would be.

File: webpack/hashed-folder-copy-plugin/src/HashedFolderCopyPlugin.ts [hunk 2]
```diff
@@ -267,21 +267,19 @@ export class HashedFolderCopyPlugin implements webpack.Plugin {
...
+          compilation.errors.push(new Error(errorMessage));
+          return this._renderError(errorMessage);
+        }
 
-          if (assetsToAdd.has(globResult)) {
-            const errorMessage: string = `Two files resolve to the same output path "${globResult}"`;
-            compilation.errors.push(new Error(errorMessage));
-            return this._renderError(errorMessage);
-          }
+        const globResultFullPath: string = path.resolve(resolvedGlobsBase, globResult);
```
Comment (D4N14L) [inline webpack/hashed-folder-copy-plugin/src/HashedFolderCopyPlugin.ts:278]: Should we have the globber return absolute paths so we don't have to do this all the time?

File: libraries/rush-lib/src/logic/buildCache/getHashesForGlobsAsync.ts [hunk 3]
```diff
@@ -13,16 +13,16 @@ async function expandGlobPatternsAsync(
 ): Promise<string[]> {
   const allMatches: Set<string> = new Set<string>();
 
-  const { default: glob } = await import('glob');
-  const globAsync = (pattern: string, options: IOptions = {}): Promise<string[]> => {
-    return LegacyAdapters.convertCallbackToPromise(glob, pattern, options);
+  const { default: glob } = await import('fast-glob');
+  const globAsync = (pattern: string, options: Options = {}): Promise<string[]> => {
```
Comment (D4N14L) [inline libraries/rush-lib/src/logic/buildCache/getHashesForGlobsAsync.ts:17]: Don't think this is needed anymore IMO

**Review by D4N14L — APPROVED** [review]

File: libraries/typings-generator/src/TypingsGenerator.ts (hunk 1 above)
Comment (dmichon-msft) [reply libraries/typings-generator/src/TypingsGenerator.ts:138]: This is really just a case of poor typings. I expect `ReadonlyArray<string>` is probably the correct input type.

File: webpack/hashed-folder-copy-plugin/src/HashedFolderCopyPlugin.ts (hunk 2 above)
Comment (dmichon-msft) [reply webpack/hashed-folder-copy-plugin/src/HashedFolderCopyPlugin.ts:278]: The relative path is used separately in this function, so in this scenario it is best to defer the resolving.

File: libraries/rush-lib/src/logic/buildCache/getHashesForGlobsAsync.ts (hunk 3 above)
Comment (dmichon-msft) [reply libraries/rush-lib/src/logic/buildCache/getHashesForGlobsAsync.ts]: I want to kill this function but haven't yet. That'll happen when all cache hashes get computed in advance of execution.

File: libraries/rush-lib/src/logic/buildCache/getHashesForGlobsAsync.ts [hunk 4]
```diff
@@ -1,55 +1,40 @@
...
+  });
+
+  if (matches.length === 0) {
     throw new Error(
       `Couldn't find any files matching provided glob patterns: ["${Array.from(globPatterns).join('", "')}"].`
     );
   }
 
-  return Array.from(allMatches);
+  return Array.from(matches);
```
Comment (D4N14L) [inline libraries/rush-lib/src/logic/buildCache/getHashesForGlobsAsync.ts:37]: Shouldn't need to allocate a new array for this

**Review by D4N14L — APPROVED** [review]

File: libraries/typings-generator/src/TypingsGenerator.ts (hunk 1 above)
Comment (iclanton) [reply libraries/typings-generator/src/TypingsGenerator.ts:138]: Should we then change the signature on our side?

**Review by D4N14L — APPROVED** [review]

**Pull request merged**